### PR TITLE
getting currentDirectory of startupProject and operating in that directo...

### PR DESCRIPTION
**Pull request by lukewaters**
...ry part of #823

### CODE REVIEW

File: src/EntityFramework.Commands/Program.cs [hunk 1]
```diff
@@ -36,13 +38,20 @@ public Program([NotNull] IApplicationEnvironment appEnv)
...
+            {
+                _startupProjectDir = Path.GetDirectoryName(libraryPath);
+            }
         }
 
         public virtual int Main([NotNull] string[] args)
         {
             Check.NotNull(args, "args");
 
-            // TODO: Enable subcommands in help
```
Comment (lukewaters) [inline src/EntityFramework.Commands/Program.cs]: fixed in 128aabd2c

File: src/EntityFramework.Commands/Program.cs [hunk 2]
```diff
@@ -36,13 +38,20 @@ public Program([NotNull] IApplicationEnvironment appEnv)
             var assemblyName = new AssemblyName(appEnv.ApplicationName);
             var assembly = Assembly.Load(assemblyName);
             _migrationTool = new MigrationTool(loggerProvider, assembly);
+
+            var library = libraryManager.GetLibraryInformation(assembly.GetName().Name);
```
Comment (bricelam) [inline src/EntityFramework.Commands/Program.cs]: `assembly` is the current project. We need to add a `-startUpProject` parameter to each command that executes user code and set the current directory based on that.

File: src/EntityFramework.Commands/Program.cs (hunk 1 above)
Comment (bricelam) [inline src/EntityFramework.Commands/Program.cs]: No, this comment was about enabling `ef help migration add` which currently throws.

Comment (lukewaters) [inline src/EntityFramework.Commands/Program.cs]: ```ef migration add help``` works, but you have a point, I’ll put the comment back in

File: src/EntityFramework.Commands/Program.cs [hunk 3]
```diff
@@ -263,6 +309,19 @@ public virtual int ShowHelp([CanBeNull] string command)
 
             return 0;
         }
+
+        public virtual string GetProjectPath(string projectName)
```
Comment (bricelam) [inline src/EntityFramework.Commands/Program.cs]: `projectName` may be `null` if the user didn't specify one. `GetLibraryInformation` could throw.

File: src/EntityFramework.Commands/Program.cs [hunk 4]
```diff
@@ -36,13 +38,14 @@ public Program([NotNull] IApplicationEnvironment appEnv)
...
             _migrationTool = new MigrationTool(loggerProvider, assembly);
+            _libraryManager = libraryManager;
         }
 
         public virtual int Main([NotNull] string[] args)
         {
             Check.NotNull(args, "args");
 
-            // TODO: Enable subcommands in help
+            var returnDirectory = Directory.GetCurrentDirectory();
```
Comment (bricelam) [inline src/EntityFramework.Commands/Program.cs]: Not used

File: src/EntityFramework.Commands/Program.cs (hunk 1 above)
Comment (bricelam) [inline src/EntityFramework.Commands/Program.cs]: Please re-add

**Comment by bricelam** [comment]
Can probably repeat yourself less if you introduce a pattern like this...
```C#
return Execute(
    startupProject,
    () =>
    {
        DoStuff();

        return 0;
    });
```
...and move the directory changing to the `Execute` method.

File: src/EntityFramework.Commands/Program.cs [hunk 5]
```diff
@@ -43,6 +46,7 @@ public virtual int Main([NotNull] string[] args)
             Check.NotNull(args, "args");
 
             // TODO: Enable subcommands in help
+            var returnDirectory = Directory.GetCurrentDirectory();
```
Comment (bricelam) [inline src/EntityFramework.Commands/Program.cs]: Not used

File: src/EntityFramework.Commands/Program.cs [hunk 6]
```diff
@@ -191,15 +209,34 @@ public virtual int AddMigration([CanBeNull] string name, [CanBeNull] string cont
...
+            var startupProjectDir = GetProjectPath(startupProject);
+            if (startupProjectDir != null)
+            {
+                Directory.SetCurrentDirectory(startupProjectDir);
+            }
+
             _migrationTool.AddMigration(name, context, _rootNamespace, _projectDir).ToArray();
 
+            Directory.SetCurrentDirectory(returnDirectory);
```
Comment (bricelam) [inline src/EntityFramework.Commands/Program.cs]: Should probably do this in a `finally` so it gets reset even when the `_migrationTool` throws.

**Comment by lukewaters** [comment]
Ping!

**Comment by bricelam** [comment]
:shipit:

**Pull request closed without merging**